paasta_tools/smartstack_tools.py

```diff
@@ -432,6 +433,41 @@ def match_backends_and_tasks(
     return backend_task_pairs
 
 
+def match_backends_and_pods(
+    backends: Iterable[HaproxyBackend], pods: Iterable[V1Pod]
+) -> List[Tuple[Optional[HaproxyBackend], Optional[V1Pod]]]:
+    """Returns tuples of matching (backend, pod) pairs, as matched by IP. Each backend will be listed exactly
+    once. If a backend does not match with a pod, (backend, None) will be included.
+    If a pod's IP does not match with any backends, (None, pod) will be included.
+
+    :param backends: An iterable of haproxy backend dictionaries, e.g. the list returned by
+                     smartstack_tools.get_multiple_backends.
+    :param pods: An iterable of V1Pod objects.
+    """
+
+    # { ip : [backend1, backend2], ... }
+    backends_by_ip: DefaultDict[str, List[HaproxyBackend]] = collections.defaultdict(
+        list
+    )
+    backend_pod_pairs = []
+
+    for backend in backends:
+        ip, port, _ = ip_port_hostname_from_svname(backend["svname"])
+        backends_by_ip[ip].append(backend)
+
+    for pod in pods:
+        ip = pod.status.pod_ip
+        for backend in backends_by_ip.pop(ip, [None]):
+            backend_pod_pairs.append((backend, pod))
```

Comment:
I don't really know my kubernetes yet and will defer to @EmanekaT, but do we expect more than one backend per pod?

Reply:
Yes. One backend per pod since a pod only runs a single instance.

Reply:
It should be one backend per pod now, but things would change after [PAASTA-14818](https://jira.yelpcorp.com/browse/PAASTA-14818). At the end, I just followed what we do for marathon for now.

Reply:
@chlgit I'm not very familiar with this. But I think those non-smartstack pods should not be exposed to synapse so they are not supposed to be backends? I guess we had ip to backend map on marathon because in Marathon world, services use node ip. But now they use pod ip, which is unique per service.

Reply:
Yup, you're also right. Currently, it's per backed per now, but this might change in the future. I don't have strong opinion here. Let me know if you would like to change.

Reply:
No this seems fine, I was more just making sure I wasn't missing anything